pkg/kubernetes/watcher.go

```diff
@@ -0,0 +1,200 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog/v2"
+
+	collector_metric "github.com/sustainable-computing-io/kepler/pkg/collector/metric"
+	"github.com/sustainable-computing-io/kepler/pkg/config"
+)
+
+const (
+	informerTimeout = time.Minute
+	podResourceType = "pods"
+)
+
+var (
+	regexReplaceContainerIDPrefix = regexp.MustCompile(`.*//`)
+)
+
+type ObjListWatcher struct {
+	// Lock to syncronize the collector update with the watcher
+	Mx *sync.Mutex
+
+	k8sCli       *kubernetes.Clientset
+	ResourceKind string
+	informer     cache.SharedInformer
+	stopChannel  chan struct{}
+
+	// ContainersMetrics holds all container energy and resource usage metrics
+	ContainersMetrics *map[string]*collector_metric.ContainerMetrics
+}
+
+func newK8sClient() *kubernetes.Clientset {
+	var restConf *rest.Config
+	var err error
+	if config.KubeConfig == "" {
+		// creates the in-cluster config
+		restConf, err = rest.InClusterConfig()
+		klog.Infoln("Using in cluster k8s config")
+	} else {
+		// use the current context in kubeconfig
+		restConf, err = clientcmd.BuildConfigFromFlags("", config.KubeConfig)
+		klog.Infoln("Using out cluster k8s config: ", config.KubeConfig)
+	}
+	if err != nil {
+		klog.Infoln("%v", err)
+		return nil
+	}
+	// creates the clientset
+	clientset, err := kubernetes.NewForConfig(restConf)
+	if err != nil {
+		klog.Fatalf("%v", err)
+	}
+	return clientset
+}
+
+func NewObjListWatcher() *ObjListWatcher {
+	w := &ObjListWatcher{
+		stopChannel:  make(chan struct{}),
+		k8sCli:       newK8sClient(),
+		ResourceKind: podResourceType,
+	}
+	if w.k8sCli == nil {
+		return w
+	}
+
+	optionsModifier := func(options *metav1.ListOptions) {
+		options.FieldSelector = fmt.Sprintf("spec.nodeName=%s", collector_metric.NodeName) // to filter events per node
+	}
+	objListWatcher := cache.NewFilteredListWatchFromClient(
+		w.k8sCli.CoreV1().RESTClient(),
+		w.ResourceKind,
+		metav1.NamespaceAll,
+		optionsModifier,
+	)
+
+	w.informer = cache.NewSharedInformer(objListWatcher, nil, 0)
+	w.stopChannel = make(chan struct{})
+	w.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(oldObj interface{}, newObj interface{}) {
+			w.handleUpdate(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			w.handleDeleted(obj)
+		},
+	})
+	return w
+}
+
+func (w *ObjListWatcher) Run() {
+	if w.k8sCli == nil {
+		return
+	}
+	go w.informer.Run(w.stopChannel)
+	timeoutCh := make(chan struct{})
+	timeoutTimer := time.AfterFunc(informerTimeout, func() {
+		close(timeoutCh)
+	})
+	defer timeoutTimer.Stop()
+	if !cache.WaitForCacheSync(timeoutCh, w.informer.HasSynced) {
+		klog.Fatalf("watcher timed out waiting for caches to sync")
+	}
+}
+
+func (w *ObjListWatcher) Stop() {
+	close(w.stopChannel)
+}
+
+func (w *ObjListWatcher) handleUpdate(obj interface{}) {
+	switch w.ResourceKind {
+	case "pods":
+		pod, ok := obj.(*k8sv1.Pod)
+		if !ok {
+			klog.Infof("Could not convert obj: %v", w.ResourceKind)
```

Comment:
I think we got code this from somewhere ? seems not able to convert should not continue following code?

Reply:
Yes, good catch!